Sources/Testing/Traits/SerialTrait.swift

```diff
@@ -0,0 +1,58 @@
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for Swift project authors
+//
+
+/// A type that affects whether or not a test or suite is parallelized.
+///
+/// When added to a parameterized test function, this trait causes that test to
+/// run its cases serially instead of in parallel. When applied to a
+/// non-parameterized test function, this trait has no effect. When applied to a
+/// test suite, this trait causes that suite to run its contained test functions
+/// and sub-suites serially instead of in parallel.
+///
+/// This trait is recursively applied: if it is applied to a suite, any
+/// parameterized tests or test suites contained in that suite are also
+/// serialized (as are any tests contained in those suites, and so on.)
+///
+/// This trait does not affect the execution of a test relative to its peers or
+/// to unrelated tests. This trait has no effect if test parallelization is
+/// globally disabled (by, for example, passing `--no-parallel` to the
+/// `swift test` command.)
+///
+/// To add this trait to a test, use ``Trait/serial``.
+public struct SerialTrait: TestTrait, SuiteTrait {
+  public var isRecursive: Bool {
+    true
+  }
+}
+
+// MARK: - SPIAwareTrait
+
+@_spi(ExperimentalTestRunning)
+extension SerialTrait: SPIAwareTrait {
+  public func prepare(for test: Test, action: inout Runner.Plan.Action) async throws {
+    if case var .run(options) = action {
+      options.isParallelizationEnabled = false
+      action = .run(options: options)
+    }
+  }
+}
+
+// MARK: -
+
+extension Trait where Self == SerialTrait {
+  /// A trait that serializes the test to which it is applied.
+  ///
+  /// ## See Also
+  ///
+  /// - ``SerialTrait``
+  public static var serial: Self {
```

Comment:
My primary comment about this PR (which introduces some _really_ useful new functionality!) is the way the trait is exposed as API. I strongly suspect that over time, we are going to want to offer more than one way to control parallelization with varying levels of granularity/scope. A few examples:

- Some test codebases might be configured to disable parallelization _by default_, but allow selected tests to opt-in to parallelization. This kind of workflow would allow incremental adoption of parallelization over time. 
- Some common testing workflows involve running more than one test target, and the supervising tool may choose to run those targets in parallel with each other. From the testing library's perspective, it may only be aware of a single process—the one running the current test target—so that is all it can _directly_ control. But a test author might need to tell the supervising tool whether it is safe to parallelize other tests in other processes at the same time as this process. If tests touch system-global state, such as the file system, this can be important.
- A user might want the test functions in a suite to _not_ be parallelized relative to each other, but might want the test cases of each of any parameterized test functions to still be parallelized with respect to the other cases — or vice-versa.

As I've been considering how we would design APIs for these use cases eventually, it feels like we'd want a slightly more generic trait which allows specifying a parallelization "style" or strategy. The above examples _could_ look something like this:

- `.parallelized(true)`: Basic opt-in to parallelization.
- `.parallelized(.serial)`: Conceptually the same as `.serial` in this PR.
- `.parallelized(.exclusivelySerial)`: A stronger form of serialization, which expands the scope to all test targets/processes.
- `.parallelized(.byCase)`: Enable parallellization of test cases, but not of test functions.

These are just examples, and the exact spellings of these things are made up for this discussion. But my thinking is that it may serve us well to provide a more generalized trait like `.parallelized` (or a different spelling) which allows _configuring_ parallelization, instead of having `SerialTrait` be a top-level trait of its own.